Core/include/Acts/Vertexing/SeedVertexFinder.ipp

```diff
@@ -0,0 +1,656 @@
+                  Acts::ActsScalar phiB = middle_sp.second;
+                  Acts::ActsScalar delta_phiAB =
+                      detail::difference_periodic(phiA, phiB, 2 * M_PI);
+                  if (std::abs(delta_phiAB) > m_cfg.maxPhideviation) {
+                    continue;
+                  }
+                  // for all far spacepoints in this slice
+                  for (const auto& far_sp :
+                       sorted_spacepoints.at(2).at(farphi).at(farz)) {
+                    Acts::ActsScalar phiC = far_sp.second;
+                    Acts::ActsScalar delta_phiBC =
+                        detail::difference_periodic(phiB, phiC, 2 * M_PI);
+                    if (std::abs(delta_phiBC) > m_cfg.maxPhideviation) {
+                      continue;
+                    }
+
+                    Acts::SeedVertexFinder<spacepoint_t>::Triplet tr(
+                        *near_sp.first, *middle_sp.first, *far_sp.first);
+
+                    if (isTripletValid(tr)) {
+                      triplets.push_back(tr);
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  return triplets;
+}
+
+template <typename spacepoint_t>
+bool Acts::SeedVertexFinder<spacepoint_t>::isTripletValid(
+    Acts::SeedVertexFinder<spacepoint_t>::Triplet& triplet) {
+  // slope for near+middle spacepoints
+  Acts::ActsScalar alpha1 =
+      std::atan2(triplet.a.y() - triplet.b.y(), triplet.a.x() - triplet.b.x());
+  // slope for middle+far spacepoints
+  Acts::ActsScalar alpha2 =
+      std::atan2(triplet.b.y() - triplet.c.y(), triplet.b.x() - triplet.c.x());
+  // these two slopes shouldn't be too different
+  Acts::ActsScalar delta_alpha =
+      detail::difference_periodic(alpha1, alpha2, 2 * M_PI);
+  if (std::abs(delta_alpha) > m_cfg.maxXYdeviation) {
+    return false;
+  }
+
+  // near-middle ray
+  Acts::Vector3 ab{triplet.a.x() - triplet.b.x(), triplet.a.y() - triplet.b.y(),
+                   triplet.a.z() - triplet.b.z()};
+  // middle-far ray
+  Acts::Vector3 bc{triplet.b.x() - triplet.c.x(), triplet.b.y() - triplet.c.y(),
+                   triplet.b.z() - triplet.c.z()};
+  // dot product of these two
+  Acts::ActsScalar costheta =
+      (ab.dot(bc)) / (std::sqrt(ab.dot(ab)) * std::sqrt(bc.dot(bc)));
+  Acts::ActsScalar theta = std::acos(costheta);
+  if (theta > m_cfg.maxXYZdeviation) {
+    return false;
+  }
+
+  // reject the ray if it doesn't come close to the z-axis
+  Acts::Ray3D ray = makeRayFromTriplet(triplet);
+  const Acts::Vector3& start_point = ray.origin();
+  const Acts::Vector3& direction = ray.dir();
+
+  Acts::ActsScalar tanTheta =
+      std::sqrt(direction[0] * direction[0] + direction[1] * direction[1]) /
+      direction[2];
+  if (std::fabs(tanTheta) < std::tan(m_cfg.minTheta)) {
+    return false;
+  }
+
+  // norm to z-axis and to the ray
+  Acts::Vector3 norm{-1. * direction[1], 1. * direction[0], 0};
```

Comment:
is this normalized? if not maybe we should normalize it and then remove the `norm.dot(norm)` below

Reply:
Hi @andiwand, we can't normalize it because `zdist` needs its square and it relies on the fact that `norm` is a cross product of `direction` and `(0,0,1)`, while both `direction` and `(0,0,1)` are normalized.
On the other hand, it seems beneficial to save size of this vector and use it instead of `norm.dot(norm)`, so I modified it in that direction.

Cheers, Petr